DS-2704

jrwishart opened the pull request.

jrwishart committed:
Add automated outlier removal in Linear Regression

Added new parameter to input the desired proportion of outliers to remove from the data and recompute the OLS. The model with the reduced data is returned as the original model. Added info in the last part of the footer about the outlier percentage

jrwishart committed:
Refactor FitRegression to allow multiple calls

Refitting the data when outlier.proportion is set should call the same code twice instead of duplicating code for the second run. To facilitate this the subset argument of the lm, glm and survey functions will be used. This required a refactor to add a logical vector to the data

jrwishart committed:
Give expectations for all tests [revdep skip]

All tests now have an 'expect_*' statement to prevent the tests being flagged as empty in the unit tests

jrwishart committed:
Consider use of SURE package for Ordered Logit residuals [revdep skip]

jrwishart committed:
Improve tests for automated outliers in Linear case [revdep skip]

jrwishart committed:
Add to prompt warnings for unusual observations

Modify the warning prompts when unusual observations are detected that consider the automated outlier removal settings. In particular

* Prompt the user to try the automated outlier removal if they currently don't and unusual observations are detected
* Prompt the user to examine the model and consider increasing the removal settings if the data was already automatically trimmed but unusual observations are still detected

jrwishart committed:
Remove multinomial logit from supported cases

Multinomial logit uses nnet::multinom which is unsupported by sure::resids which is a viable contender for automated outlier removal using residuals as the metric. The VGAM::vglm implementation of multinomial models is supported by sure::resids. Plan to update Multinomial Logit code base in the future and update the automated outlier removal to support Multinomial Logit

jrwishart committed:
Add random seed for Surrogate residuals

Surrogate residuals are not deterministic and depend on RNG. Seed parameter is set in the relevant functions and it is shared between the current seed parameter used for imputation

jrwishart committed:
Update unit tests to include weighted cases

jrwishart committed:
Implement changes to handle syvolr

jrwishart committed:
Update tests for all regression cases

jrwishart committed:
Update argument name and improve tests

Updated argument name from outlier.proportion to outlier.proportion.to.remove to make it more verbose and clear about its function. Also added unit tests to check the proportion is adhered to for each test block.

jrwishart committed:
Update relative importance and shapley

Update code for importance analysis to account for the refitting and filter the data appropriately. Also added unit tests to cover the linear case for importance analysis and shapley

jrwishart committed:
Update Relative Importance Analysis and Shapley across all regression

Modified the logic to ensure the appropriate metrics, R squared and the matrix calculations for the two methods are computed on the appropriate subsets of the data when outliers have been removed. Updated unit tests to have all coverage of Regression types in this case

jrwishart committed:
Fix formula to handle dot operator and outlier indicator variable

jrwishart committed:
Update footer for RIA and Shapley

mwmclean commented on R/regression.R:
```diff
@@ -618,46 +637,102 @@ importanceFooter <- function(x)
+    {
+        # Don't support Multinomial Logit for now.
+        if (type == "Multinomial Logit")
+            stop("Automated outlier removal and re-fitting a 'Multinomial Logit' model is not supported")
```
what about instead warning and continuing without the outlier removal in the multinomial case?

jrwishart replied:
I agree that is better. Code amended to do this in https://github.com/Displayr/flipRegression/commit/4d50650a1ac03dd71c97b804f034445ad1e0bedc

mwmclean commented on tests/testthat/test-_bugs.R:
```diff
@@ -86,12 +86,12 @@ test_that("Duplicated labels",
         20L), class = "data.frame")
-
-    expect_warning(print(Regression(Q4_2~.,, data = df.small, show.labels = TRUE,
+    test.formula <- as.formula(paste0("Q4_2 ~ ", paste0(names(df.small)[-1], collapse = " + ")))
```
Why was the `.` removed from these tests?

jrwishart replied:
I don't recall exactly why I did that for this case. I did go through a lot of test files and tidied up some things. It doesn't look necessary in this case and the tests still pass when the `.` is restored in the formula. Code restored in https://github.com/Displayr/flipRegression/commit/4d50650a1ac03dd71c97b804f034445ad1e0bedc

mwmclean commented on tests/testthat/test-multipleimputation.R:
```diff
@@ -43,90 +43,92 @@ test_that("Multiple imputation with auxiliary variables ", {
     est <- flipData::EstimationData(Overall ~ Fees + Interest + Phone + Branch + Online + ATM + rnd + rnd1 + rnd2 + rnd3 + rnd4, zbank, missing = "Multiple imputation", m = 10)
     models <- lapply(est$estimation.data, FUN = function(x) Regression(Overall ~ Fees + Interest + Phone + Branch + Online + ATM + rnd + rnd1 + rnd2 + rnd3 + rnd4, data = x))
-    coefs <- MultipleImputationCoefficientTable(models)
+    coefs <- expect_error(MultipleImputationCoefficientTable(models), NA)
```
can remove `coefs <-`

jrwishart replied:
Removed in https://github.com/Displayr/flipRegression/commit/4d50650a1ac03dd71c97b804f034445ad1e0bedc

mwmclean commented on R/regression.R:
```diff
@@ -62,6 +64,10 @@
+#'   denote that a percentage between none and 50\% of the data points are removed.
 #' @param ... Additional argments to be passed to  \code{\link{lm}} or, if the
 #'  data is weighted,  \code{\link[survey]{svyglm}} or \code{\link[survey]{svyolr}}.
 #' @details In the case of Ordered Logistic regression, this function computes a proporional odds model using
```
Would be good to document how the outlier detection works in the details and discussed the types of residuals used in the various cases.

jrwishart replied:
Added more details in the documentation in https://github.com/Displayr/flipRegression/commit/4d50650a1ac03dd71c97b804f034445ad1e0bedc. Let me know if more details are required.

mwmclean reviewed (commented):
![image](https://user-images.githubusercontent.com/5276419/74218748-e4aa6d80-4cfe-11ea-9c0f-ffc2bd791f2d.png)
Unfortunately, seeing a bunch of errors on my machine when running locally.

jrwishart committed:
Update documentation and address code review comments

Updated the documentation in Regression to give details and references
for the residual calculations used in all the supported Regression
types.

Also addressed comments in code review for DS-2704, including changing
error to warning for Multinomial Logit in automated outlier removal and
small test updates.

jrwishart commented:
Locally I still get all tests passing and the travis tests pass too. The only discrepancy I have is an RNG issue from testing the entire package (i.e. `devtools::test()`) results in 8 errors in https://github.com/Displayr/flipRegression/blob/750ae2e99b8b8643d1c55d65da620377beff21d0/tests/testthat/test-regression.R#L25-L48. However those same lines pass when the file is tested individually (i.e. `devtools::test_file("tests/testthat/test-regression.R")`. Those lines were changed in https://github.com/Displayr/flipRegression/commit/59d53685eeb09fed29d0e6a8fd20b64695c0c390 Those 8 tests pass on travis.
